nav2_collision_monitor/src/collision_monitor_node.cpp

```diff
@@ -0,0 +1,617 @@
+
+        p->configure();
+
+        sources_.push_back(p);
+      } else {  // Error if something else
+        RCLCPP_ERROR(
+          get_logger(),
+          "[%s]: Unknown source type: %s",
+          source_name.c_str(), source_type.c_str());
+        return false;
+      }
+    }
+  } catch (const std::exception & ex) {
+    RCLCPP_ERROR(get_logger(), "Error while getting parameters: %s", ex.what());
+    return false;
+  }
+
+  return true;
+}
+
+void CollisionMonitor::process(const Velocity & cmd_vel_in)
+{
+  // Current timestamp for all inner routines prolongation
+  rclcpp::Time curr_time = this->now();
+
+  // Do nothing if main worker in non-active state
+  if (!process_active_) {
+    return;
+  }
+
+  // Warn if odometry was not received yet and do nothing
+  if (odom_frame_id_.length() == 0) {
+    RCLCPP_WARN(get_logger(), "Odometry is not received yet");
+    return;
+  }
+
+  // Points array collected from different data sources in a robot base frame
+  std::vector<Point> collision_points;
+
+  // Fill collision_points array from different data sources
+  for (std::shared_ptr<Source> source : sources_) {
+    source->getData(
+      base_frame_id_, curr_time, odom_frame_id_, transform_tolerance_,
+      data_timeout_, tf_buffer_, collision_points);
+  }
+
+  // By default - there is no action
+  Action robot_action{DO_NOTHING, cmd_vel_in};
+  // Polygon causing robot action (if any)
+  std::shared_ptr<Polygon> action_polygon;
+
+  // Process STOP and SLOWDOWN polygons
+  for (std::shared_ptr<Polygon> polygon : polygons_) {
+    if (robot_action.action_type == STOP) {
+      // If robot already should stop, do nothing
+      break;
+    }
+
+    // Process selected polygon
+    if (processStopSlowdown(polygon, collision_points, cmd_vel_in, robot_action)) {
+      action_polygon = polygon;
+    }
+  }
+
+  // Process APPROACH footprint
+  if (approach_ && robot_action.action_type != STOP) {
+    if (processApproach(collision_points, cmd_vel_in, robot_action)) {
+      action_polygon = footprint_;
+    }
+  }
+
+  if (robot_action.action_type != action_type_prev_) {
+    // Report changed robot behavior
+    printAction(robot_action, action_polygon);
+    action_type_prev_ = robot_action.action_type;
+  }
+
+  // Publish requred robot velocity
+  publishVelocity(robot_action.req_vel);
+
+  // Publish polygons for better visualization
+  publishPolygons();
+}
+
+bool CollisionMonitor::processStopSlowdown(
+  const std::shared_ptr<Polygon> polygon,
+  const std::vector<Point> & collision_points,
+  const Velocity & velocity,
+  Action & robot_action) const
+{
+  if (polygon->getPointsInside(collision_points) > polygon->getMaxPoints()) {
+    if (polygon->getActionType() == STOP) {
+      // Setting up zero velocity for STOP model
+      robot_action.action_type = STOP;
+      robot_action.req_vel.x = 0.0;
+      robot_action.req_vel.y = 0.0;
+      robot_action.req_vel.tw = 0.0;
+      return true;
+    } else {  // SLOWDOWN
```

Comment:
That is not true, what if the polygon is an approach polygon?

Reply:
This is being checked one layer above, when `processStopSlowdown()` is being called:
```
 if (at == STOP || at == SLOWDOWN) {
      // Process STOP/SLOWDOWN for the selected polygon
      if (processStopSlowdown(polygon, collision_points, cmd_vel_in, robot_action)) {
        action_polygon = polygon;
      }
```